### PR TITLE
Refactor Item Hooks

### PR DESCRIPTION
**Pull request by Oribuin**
Remove convolution when adding new external plugin item types. This change aims to improve how plugins handle item hooks to be friendlier

### CODE REVIEW

File: src/main/java/com/extendedclip/deluxemenus/config/DeluxeMenusConfig.java [hunk 1]
```diff
@@ -57,11 +59,12 @@ public class DeluxeMenusConfig {
...
-    VALID_MATERIAL_PREFIXES.add(PLACEHOLDER_PREFIX);
-    VALID_MATERIAL_PREFIXES.add(ITEMSADDER_PREFIX);
-    VALID_MATERIAL_PREFIXES.add(ORAXEN_PREFIX);
-    VALID_MATERIAL_PREFIXES.add(MMOITEMS_PREFIX);
+    VALID_MATERIAL_PREFIXES.addAll(
+        DeluxeMenus.getInstance().getItemHooks().values()
+            .stream()
+            .map(ItemHook::getPrefix)
+            .collect(Collectors.toList())
+    );
```
Comment (Kqliber) [inline src/main/java/com/extendedclip/deluxemenus/config/DeluxeMenusConfig.java:67]: does this work? surely DeluxeMenus hasnt initialized before the static block is ran and definitely not the item hooks?

File: src/main/java/com/extendedclip/deluxemenus/menu/MenuItem.java [hunk 2]
```diff
@@ -74,31 +75,33 @@ public ItemStack getItemStack(@NotNull final MenuHolder holder) {
...
+        final String finalMaterial = lowercaseStringMaterial;
+        final ItemHook pluginHook = DeluxeMenus.getInstance().getItemHooks().values()
+            .stream()
+            .filter(x -> finalMaterial.startsWith(x.getPrefix()))
+            .findFirst()
+            .orElse(null);
+
+        if (pluginHook != null) {
+            itemStack = pluginHook.getItem(stringMaterial.substring(pluginHook.getPrefix().length()));
+        }
```
Comment (Kqliber) [inline src/main/java/com/extendedclip/deluxemenus/menu/MenuItem.java:87]: orElse and if statement can be simplified with findFirst().ifPresent()

**Review by Kqliber — COMMENTED** [review]
will approve if that static block for VALID_MATERIAL_PREFIXES has been tested and works

File: src/main/java/com/extendedclip/deluxemenus/config/DeluxeMenusConfig.java (hunk 1 above)
Comment (Oribuin) [reply src/main/java/com/extendedclip/deluxemenus/config/DeluxeMenusConfig.java:67]: I had to shuffle around when the plugin setups item hooks to be initialized before the config is to allow this to work.

image
image

File: src/main/java/com/extendedclip/deluxemenus/menu/MenuItem.java (hunk 2 above)
Comment (Oribuin) [reply src/main/java/com/extendedclip/deluxemenus/menu/MenuItem.java:87]: When using ifPresent, we would have to turn ItemStack into an atomic for it to function

**Review by lukemango — APPROVED** [review]
lgtm

File: src/main/java/com/extendedclip/deluxemenus/DeluxeMenus.java [hunk 3]
```diff
@@ -102,6 +102,20 @@ public void onEnable() {
 
     this.adventure = BukkitAudiences.create(this);
 
+    setupItemHooks();
+
+    if (Bukkit.getPluginManager().getPlugin("Vault") != null) {
```
Comment (iGabyTM) [inline src/main/java/com/extendedclip/deluxemenus/DeluxeMenus.java:107]: I think `isPluginEnabled()` is better to check if a plugin is installed **and** enabled.

**Review by iGabyTM — CHANGES_REQUESTED** [review]

**Review by Kqliber — APPROVED** [review]

File: src/main/java/com/extendedclip/deluxemenus/DeluxeMenus.java (hunk 3 above)
Comment (Oribuin) [reply src/main/java/com/extendedclip/deluxemenus/DeluxeMenus.java]: Done

**Review by iGabyTM — APPROVED** [review]

**Review by darbyjack — APPROVED** [review]

**Pull request merged**